### PR TITLE
Minor standard library constification

### PR DESCRIPTION
**Pull request by Centril**
This PR makes some bits of the standard library into `const fn`s.
I've tried to be as aggressive as I possibly could in the constification.
The list is rather small due to how restrictive `const fn` is at the moment.

r? @oli-obk cc @rust-lang/libs 

Stable public APIs affected:
+ [x] `Cell::as_ptr`
+ [x] `UnsafeCell::get`
+ [x] `char::is_ascii`
+ [x] `iter::empty`
+ [x] `ManuallyDrop::{new, into_inner}`
+ [x] `RangeInclusive::{start, end}`
+ [x] `NonNull::as_ptr`
+ [x] `{[T], str}::as_ptr`
+ [x] `Duration::{as_secs, subsec_millis, subsec_micros, subsec_nanos}`
+ [x] `CStr::as_ptr`
+ [x] `Ipv4Addr::is_unspecified`
+ [x] `Ipv6Addr::new`
+ [x] `Ipv6Addr::octets`

Unstable public APIs affected:
+ [x] `Duration::{as_millis, as_micros, as_nanos, as_float_secs}`
+ [x] `Wrapping::{count_ones, count_zeros, trailing_zeros, rotate_left, rotate_right, swap_bytes, reverse_bits, from_be, from_le, to_be, to_le, leading_zeros, is_positive, is_negative, leading_zeros}`
+ [x] `core::convert::identity`

--------------------------

## Removed from list in first pass:

Stable public APIs affected:
+ [ ] `BTree{Map, Set}::{len, is_empty}`
+ [ ] `VecDeque::is_empty`
+ [ ] `String::{is_empty, len}`
+ [ ] `FromUtf8Error::utf8_error`
+ [ ] `Vec<T>::{is_empty, len}`
+ [ ] `Layout::size`
+ [ ] `DecodeUtf16Error::unpaired_surrogate`
+ [ ] `core::fmt::{fill, width, precision, sign_plus, sign_minus, alternate, sign_aware_zero_pad}`
+ [ ] `panic::Location::{file, line, column}`
+ [ ] `{ChunksExact, RChunksExact}::remainder`
+ [ ] `Utf8Error::valid_up_to`
+ [ ] `VacantEntry::key`
+ [ ] `NulError::nul_position`
+ [ ] `IntoStringError::utf8_error`
+ [ ] `IntoInnerError::error`
+ [ ] `io::Chain::get_ref`
+ [ ] `io::Take::{limit, get_ref}`
+ [ ] `SocketAddrV6::{flowinfo, scope_id}`
+ [ ] `PrefixComponent::{kind, as_os_str}`
+ [ ] `Path::{ancestors, display}`
+ [ ] `WaitTimeoutResult::timed_out`
+ [ ] `Receiver::{iter, try_iter}`
+ [ ] `thread::JoinHandle::thread`
+ [ ] `SystemTimeError::duration`

Unstable public APIs affected:
+ [ ] `core::fmt::Arguments::new_v1`
+ [ ] `core::fmt::Arguments::new_v1_formatted`
+ [ ] `Pin::{get_ref, into_ref}`
+ [ ] `Utf8Lossy::chunks`
+ [ ] `LocalWaker::as_waker`
+ [ ] `panic::PanicInfo::{internal_constructor, message, location}`
+ [ ] `panic::Location::{internal_constructor }`

## Removed from list in 2nd pass:

Stable public APIs affected:
+ [ ] `LinkedList::{new, iter, is_empty, len}`
+ [ ] `mem::forget`
+ [ ] `Cursor::{new, get_ref, position}`
+ [ ] `io::{empty, repeat, sink}`
+ [ ] `PoisonError::new`
+ [ ] `thread::Builder::new`
+ [ ] `process::Stdio::{piped, inherit, null}`

Unstable public APIs affected:
+ [ ] `io::Initializer::{zeroing, should_initialize}`

### CODE REVIEW

**Comment by sfackler** [comment]
Many of these are literally impossible to use in a const context. Why are we doing this?

**Comment by Centril** [comment]
+ I wanted to see how much we could possible do right now -- I fully expect you'll be wanting to reduce the list :)
+ Some of the ones, like `vec.is_empty()` that you can't use in a const context right now are due to the lack of `const fn` constructors. Once you get those, these will be usable in const contexts. We could wait until then, but this felt easier to do it at once instead of people asking for one after the other. I.e. asking "has this become useful in const contexts" each time on every function seems too time consuming.

   The question to be asked on each function I think is: "is there some reason reason why we are not comfortable guaranteeing that this is `const fn` because we might want to introduce non-const behavior later?"
+ `const fn` is not just about being usable in const contexts; it's also saying "this is referentially transparent".

**Review by oli-obk — COMMENTED** [review]
Most of these seem fine in isolation.

I don't think it is good practice to make functions const fn while they have an argument that is not constructable in a const context.

Doing so will incentivise people to use unions or transmute to construct values in const contexts.

**Comment by withoutboats** [comment]
> I don't think it is good practice to make functions const fn while they have an argument that is not constructable in a const context.

Agreed.

Right now we haven't established a shared vision about what should be const. There's an argument to be made that anything that can be const should be, but that argument has not been made yet.

I think as a first pass, to make this list more manageable, I'd like to see every function removed that can't actually be called in a const context already.

> const fn is not just about being usable in const contexts; it's also saying "this is referentially transparent".

const fn isn't intended to mean that the function is referentially transparent. In particular, there's not a technical reason we could not allow mutable references as arguments to const fn, which would violate any referential transparency. `const fn` is not `pure fn`.

**Comment by Centril** [comment]
> Right now we haven't established a shared vision about what should be const. There's an argument to be made that anything that can be const should be, but that argument has not been made yet.

That's my vision at least; I guess it might need to be RFCed to become shared.

> I think as a first pass, to make this list more manageable, I'd like to see every function removed that can't actually be called in a const context already.

Will do. :)

> const fn isn't intended to mean that the function is referentially transparent. In particular, there's not a technical reason we could not allow mutable references as arguments to const fn, which would violate any referential transparency. const fn is not pure fn.

Mutable references would not really violate referential transparency in my view and I have no problem allowing them in `const fn`. With a function such as `const fn foo(y: A, x: &mut B) -> C {...}`  you are in effect saying the same thing as `A -> B -> (B, C)`, which is equivalent to `A -> State B C`. Yet, state monads do not cause global side effects in Haskell. To me, `const fn` is `pure fn` in all but name.

**Comment by Centril** [comment]
> > I think as a first pass, to make this list more manageable, I'd like to see every function removed that can't actually be called in a const context already.
> 
> Will do. :)

@withoutboats Done. I've segregated the list; the ones that are removed are in their separate list in the OP.

File: src/liballoc/collections/binary_heap.rs [hunk 1]
```diff
@@ -884,7 +884,7 @@ impl<'a, T> Hole<'a, T> {
     }
 
     #[inline]
-    fn pos(&self) -> usize {
+    const fn pos(&self) -> usize {
```
Comment (oli-obk) [inline src/liballoc/collections/binary_heap.rs]: There's no const constructor for `Hole`

File: src/liballoc/collections/btree/node.rs [hunk 2]
```diff
@@ -357,7 +357,7 @@ impl<BorrowType, K, V, Type> NodeRef<BorrowType, K, V, Type> {
 
     /// Returns the height of this node in the whole tree. Zero height denotes the
     /// leaf level.
-    pub fn height(&self) -> usize {
+    pub const fn height(&self) -> usize {
```
Comment (oli-obk) [inline src/liballoc/collections/btree/node.rs]: no const constructor

File: src/liballoc/collections/vec_deque.rs [hunk 3]
```diff
@@ -1273,7 +1273,7 @@ impl<T> VecDeque<T> {
     }
 
     #[inline]
-    fn is_contiguous(&self) -> bool {
+    const fn is_contiguous(&self) -> bool {
```
Comment (oli-obk) [inline src/liballoc/collections/vec_deque.rs]: no const constructor

**Review by oli-obk — CHANGES_REQUESTED** [review]

File: src/liballoc/collections/binary_heap.rs (hunk 1 above)
Comment (Centril) [reply src/liballoc/collections/binary_heap.rs]: These are all *internal* APIs.

File: src/liballoc/collections/btree/node.rs (hunk 2 above)
Comment (Centril) [reply src/liballoc/collections/btree/node.rs]: Ditto.

File: src/liballoc/collections/vec_deque.rs (hunk 3 above)
Comment (Centril) [reply src/liballoc/collections/vec_deque.rs]: Ditto.

File: src/liballoc/collections/binary_heap.rs (hunk 1 above)
Comment (rkruppe) [reply src/liballoc/collections/binary_heap.rs]: Then why constify them at all?

Comment (Centril) [reply src/liballoc/collections/binary_heap.rs]: To increase the code quality and to make the eventual constification of the public APIs that use these internal APIs easier once we give `const fn` more powers.

Comment (rkruppe) [reply src/liballoc/collections/binary_heap.rs]: How does an annotation that currently has no benefit "increase code quality"? And with respect to future changes: if and when we make some public API that relies on this `const`, doing this one-line change then is no significant additional burden. Furthermore, until that day it's just line noise that can hinder or get lost in refactorings.

Comment (rkruppe) [reply src/liballoc/collections/binary_heap.rs]: Summarizing some discord discussion: Using `const fn` to indicate lack of side effects, divorced from CTFE, is your preference but not consensus. Even besides the question of "are `const fn`s pure?", there's no coding style decision to indicate as many pure functions as possible with `const` if they are not relevant for CTFE (one could consider that annotation burden not worthwhile). Without such a style decision, it seems at best neutral to use it that way on one's own accord.

Comment (Centril) [reply src/liballoc/collections/binary_heap.rs]: Summarizing what I said on Discord; if @oli-obk wants me to change it, I will. But it seems to me a supreme waste of time.

Comment (sfackler) [reply src/liballoc/collections/binary_heap.rs]: *Not* making a change is a supreme waste of time?

**Comment by Centril** [comment]
Removed changes to internal APIs per review.

**Review by oli-obk — APPROVED** [review]

**Comment by oli-obk** [comment]
I am for this change. But we do have a ban on creating new stable public const fns in the libstd, and we haven't officially lifted that ban.

cc @nikomatsakis I am reasonably sure that we have fixed everything that can go wrong with new public const fns. Reasons we didn't stabilize any new ones before:

* the `const_fn` feature gate is too powerful, we weren't really sure what we were exposing
    * resolved via `min_const_fn`
* const fns were immediately promotable
    * no libstd/libcore function is promotable unless marked so with the `rustc_const_promotable` attribute
    * no user defined const fns are promotable unless the `const_fn` feature gate is enabled

I propose we lift the ban on adding new const fns to the libstd

**Comment by Centril** [comment]
> But we do have a ban on creating new stable public const fns in the libstd, and we haven't officially lifted that ban.

I believe the ban was originally imposed by the language team for the reasons @oli-obk mentioned.
I think from the language team's perspective they should be considered lifted given the stabilization of min_const_fn and the changes mentioned. Indeed, stabilized specifically in such a way to facilitate stabilization of new functions in the standard library as const fn.

> I propose we lift the ban on adding new const fns to the libstd

Yes, let's :)

**Comment by alexcrichton** [comment]
The PR title currently says it's a "Minor standard library constification" but then the PR text says:

>  I've tried to be as aggressive as I possibly could in the constification.

It looks like this has been scaled back since the first round but we need to be very careful to not over-promise on implementations here. Landing this PR will require an FCP signoff from the @rust-lang/libs team.

I personally still think that the list is too ambitious and would like to see it cut back further. We can always add `const` whenever we want, it's impossible to remove it once stabilized. I'd rather we take our time and move deliberately rather than simply codify what today's implementation happens to be.

For example I don't think we should stabilize the `const`-ness of the following APIs:

* `LinkedList::iter` - the iterator returned can't be used?
* `char::is_ascii` - is anyone asking for this?
* `mem::forget` - is anyone asking for this?
* `Cursor` - I don't think this is ever used in a const context
* `io::*` - is this needed yet really?
* `Ipv4Addr::is_unspecified` - this seems... random? If `new` isn't even const why would this be?
* `PoisonError::new` - I kinda doubt anyone wants this
* `thread::Builder::new` - I *really* doubt anyone wants this
* `process::Stdio::{piped, inherit, null}` - I think this is overpromising and not getting much benefit, I think it should be removed from this list

**Comment by Centril** [comment]
> The PR title currently says it's a "Minor standard library constification" but then the PR text says:
> 
> > I've tried to be as aggressive as I possibly could in the constification.

As aggressive as possible given what stable `const fn` can do; which isn't much ;)
Once we can do control flow (`match` and friends), `let` bindings, loops, and panics in `const fn` then I think we should be many many times more aggressive than this.

> I personally still think that the list is too ambitious and would like to see it cut back further. We can always add `const` whenever we want, it's impossible to remove it once stabilized. I'd rather we take our time and move deliberately rather than simply codify what today's implementation happens to be.

Fair enough; I personally think we should ask the question "is there any reason we would introduce non-const behavior in this function making it not a `const fn`?" rather than "is anyone asking for this". As I noted before, once `const fn` gets more powerful, which it should in the coming months I think, the number of things that can be `const fn` in the standard library will increase dramatically; In such a future, being too deliberate on each function will take a long time to go through. When the power increases, I would at least recommend the libs team to take stock of everything in the standard library.

>     * `char::is_ascii` - is anyone asking for this?

In the sense that someone has explicitly asked for it? No.
The proposed list came about by looking at every single standard library function and then seeing if it could be constified.

The definition is fairly simple, so you should be able to copy it readily.
However, I think if we ask "is anyone asking for this?" for each of these kinds of functions then constification will take considerable time. So I'd personally be more eager in constification than doing it lazily when people ask for it.

>     * `Ipv4Addr::is_unspecified` - this seems... random? If `new` isn't even const why would this be?

`new` is const on nightly at least, https://doc.rust-lang.org/nightly/std/net/struct.Ipv4Addr.html#method.new so it can be used.

**Comment by RalfJung** [comment]
@withoutboats wrote
> const fn is not pure fn.

And I agree. I appreciate that is your vision, @Centril, but until we have adapted this vision as our shared vision, please do not use it to justify changes that we cannot take back later.

---

> I personally think we should ask the question "is there any reason we would introduce non-const behavior in this function making it not a const fn?" rather than "is anyone asking for this".

I think the question should be "is this useful?" -- as in, can you as the advocate of these features come up with an example where using this function as a `const fn` actually makes sense, an example that seems somewhat realistic and that compiles with your PR but does not without?

Things related to concurrency (`thread`, `PoisonError`) and IO (`io`, `process`)   immediately fail this criterion. `Cursor` is also useless because its point is the `Seek` implementation, and we don't get that in const context. 

`LinkedList::new` makes sense, we usually have `const fn` container constructors (and they have a clear purpose). Less so for the other methods -- we don't have `len` and `is_empty` be `const fn` for other containers either, do we? I think we should do this more consistently, in particular given that you cannot actually *fill* a container in `const fn` and hence these getters and iterators don't make *that* much sense. I don't like stabilizing a *different* random subset of the API that `LinkedList`, `Vec`, `VecDeque` etc all share to be `const fn` -- these APIs being so uniform is extremely useful, we shouldn't give up that uniformity without a good reason.

`mem::forget` isn't useful yet as we cannot have `Drop` types in const, can we? But I do not mind very strongly.

`is_ascii` and `is_unspecified` are fine, I think, as are all the remaining things not on @alexcrichton's list.

**Comment by Centril** [comment]
I've trimmed the list along the lines per @alexcrichton's list; I kept `is_ascii` and `is_unspecified` per @RalfJung's notes. I hope that's OK.

**Comment by nikomatsakis** [comment]
My take:

- @oli-obk convinced me we can lift the ban
- I agree with @RalfJung's criteria basically 100%

The one that I am torn about is `mem::forget`, because it feels like a "primitive capability" that we will eventually want to expose -- but it can also wait, and perhaps it's better not to send misleading signals (i.e., you can't actually use this productively in a const anyway...).

File: src/liballoc/collections/linked_list.rs [hunk 4]
```diff
@@ -402,7 +402,7 @@ impl<T> LinkedList<T> {
     #[inline]
     #[stable(feature = "rust1", since = "1.0.0")]
     pub fn is_empty(&self) -> bool {
-        self.head.is_none()
+        self.len() == 0
```
Comment (RalfJung) [inline src/liballoc/collections/linked_list.rs]: Why does this get changed here?

**Comment by alexcrichton** [comment]
Ok I think the list looks good now, but I'd like to get libs team signoff as well:

@rfcbot fcp merge

**Comment by glaebhoerl** [comment]
Maybe general matters like this are better discussed somewhere else (like an RFC or tracking issue), but, since it came up...

I think `const fn` has at least two meanings which are 'stronger' than its possible "pure fn" meaning:

 * "No extern calls" - even if we make MIRI all-powerful and make *everything* legal in `const`-context, it still cannot see into `extern` and C functions. (Even if they might in fact be pure!) This is the actual hard limitation which `const fn` statically guards.

 * "Semantically inline" - once a function is marked `const`, changing its *implementation* can affect whether other code *typechecks*. Changes of behavior are almost certainly ruled out; other changes-of-implementation may or may not be. This might not be a very practically relevant consideration *now*, since e.g. `const` generics aren't yet available/stable, `const fn` is extremely limited, and the functions which nonetheless pass that test are usually extremely simple, but I expect it will be one of the major implications of `const`ness in the future.

At the confluence of the two, calling a function marked `const` from a dynamically-linked library, and that library being independently updated in an ABI compatible way, becomes very likely unsound or impossible. (Again, we don't have ABI guarantees to speak of *right now*, but.)

After thinking it through a bit I guess I do agree that purity is *also* a meaning of `const fn`, but it feels like a consequence rather than an original purpose, and subordinate as a consideration to the other two.

(BTW - as you may have heard there is some controversy about the definition of "referentially transparent". Let's not get into those weeds here, the upshot is just that it's best if we stick to using "pure".:)

**Comment by alexcrichton** [comment]
Ok! @Centril want to rebase and I can r+?

Also as a passing read, I think that `AtomicUsize::new` and for other types should also be "easily `const`", right?

**Comment by SimonSapin** [comment]
`AtomicUsize::new` is already a `const fn`.

**Comment by alexcrichton** [comment]
Er right sorry, I was mixing up making-things-const and stabilizing-things-to-be-called-as-const

File: src/test/ui/consts/const-eval/mod-static-with-const-fn.rs [hunk 5]
```diff
@@ -25,11 +25,9 @@ static FOO: Foo = Foo(UnsafeCell::new(42));
 
 static BAR: () = unsafe {
     *FOO.0.get() = 5;
-    //~^ ERROR calls in statics are limited to constant functions, tuple structs and tuple variants
```
[the remark this answers is not in the document]
Comment (Centril) [reply src/test/ui/consts/const-eval/mod-static-with-const-fn.rs]: Fixed.

**Comment by Centril** [comment]
@alexcrichton rebased + minor other fixes. :)

**Comment by Mark-Simulacrum** [comment]
@Centril Is the top comment up-to-date? Mostly asking to help us down the line when making release notes.

**Comment by Centril** [comment]
@Mark-Simulacrum I can't say it is with confidence but it probably is; I'd search for `pub const fn` to be sure.

**Comment by memoryruins** [comment]
The top comment appears to be up-to-date.

If needed for release notes, "affected" includes two types of changes -- prepending ``const`` to ``fn`` and removing ``rustc_const_unstable`` feature attributes. The following lists where attributes were removed from preexisting ``const fn``s:

Stable public APIs affected:
``ManuallyDrop::new``
``{[T], str}::as_ptr``
``Duration::{as_secs, subsec_millis, subsec_micros, subsec_nanos}``
``Ipv6Addr::new``

Unstable public APIs affected:
``core::convert::identity``

The rest of the top comment prepends ``const`` to ``fn``.

**Pull request merged**